src/Compilers/CSharp/Portable/Utilities/ValueSetFactory.NumericValueSet.cs

```diff
@@ -243,6 +260,17 @@ private static T Max(T a, T b)
 
             public IValueSet<T> Union(IValueSet<T> o)
             {
+                // We use non-negative integers for Count/Length on types that list-patterns can be used on (ie.countable and indexable ones).
+                // But we need to upgrade them to regular integers to perform operations against full integer sets.
+                if (this is NumericValueSet<int, NonNegativeIntTC> nonNegativeThis && o is NumericValueSet<int, IntTC>)
+                {
+                    return ((IValueSet<T>)ExpandToIntegerRange(nonNegativeThis)).Union(o);
```

Comment:
>ExpandToIntegerRange [](http://example.com/codeflow?start=42&length=20)

I am not sure what is the scenario for hitting this code path (what new test needs this change?), but it is quite possible that we don't want to expand the set in this case either. We determined that the property doesn't have negative values, therefore, it doesn't seem useful to include negative values into the set of possible values, even if the other check didn't eliminate them on its own.

Reply:
We need to modify the Union logic for the following tests: `MixedCountPatterns_ReverseOrder`, `MixedCountPatterns_NegativeTestAfterRegularPropertyPattern` and `MixedCountPatterns_NegativeTest` (which otherwise fail with a cast exception as we try to union sets of different kinds).
That said, it looks like I only managed to hit the second branch (below). I'll see if I can manage to hit this first branch too.

Reply:
I've removed the handling of two mirror cases that I don't think we need after all. Thanks